### DISCUSSION TITLE
Can this provider be used to create a VM with iSCSI disk?

### DISCUSSION THREAD

**Discussion by NubemInvidia** [bpg/terraform-provider-proxmox#1677] [Q&A]
I can create VMs just fine when using local LVMs on the node. But I can't seem to use iSCSI targets from my SAN as the disks for the VMs. It fails to provision the disk, stating that it can't size the disk. Seems to me that even if I omit the disk size, it defaults to 8, but since it's an iSCSI target you can't specify the space the VM should use on that target so it errors out. 

```
terraform {
  required_providers {
    proxmox = {
      source = "bpg/proxmox"
      version = "0.68.1"
    }
  }
}
provider "proxmox" {
  endpoint = var.virtual_environment_endpoint
  api_token = var.virtual_environment_api_token
  insecure = true

  ssh {
    agent = true
    username = var.virtual_environment_ssh_username
    password = var.virtual_environment_ssh_password
  }
}

resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
  name        = "hostname"
  description = "Managed by Terraform"
  tags        = ["terraform", "ubuntu"]

  node_name = "pve_cluster_node2"
  vm_id     = 800

  agent {
    # read 'Qemu guest agent' section, change to true only when ready
    enabled = false
  }
  # if agent is not enabled, the VM may not be able to shutdown properly, and may need to be forced off
  stop_on_destroy = true

  startup {
    order      = "1"
    up_delay   = "10"
    down_delay = "10"
  }

  cpu {
    cores        = 2
    type         = "x86-64-v2-AES"  # recommended for modern CPUs
  }

  memory {
    dedicated = 2048
    floating  = 2048 # set equal to dedicated to enable ballooning
  }

  # Define the bootable ISO disk
  cdrom {
    enabled   = true
    file_id   = "local:iso/ubuntu-24.04.1-live-server-amd64.iso"
    interface = "ide0"
  }

  disk {
    datastore_id = "iscsi-test"
    interface = "scsi0"
    iothread = true
    aio = "native"
    cache = "none"
    file_format = "raw"
    size = 16
  }

  initialization {
    ip_config {
      ipv4 {
        address = "dhcp"
      }
    }

    user_account {
      keys     = [trimspace(tls_private_key.ubuntu_vm_key.public_key_openssh)]
      password = random_password.ubuntu_vm_password.result
      username = "ubuntu"
    }

  }

  network_device {
    bridge = "vmbr0"
  }

  operating_system {
    type = "l26"
  }

  tpm_state {
    version = "v2.0"
  }

  serial_device {}
}

resource "random_password" "ubuntu_vm_password" {
  length           = 16
  override_special = "_%@"
  special          = true
}

resource "tls_private_key" "ubuntu_vm_key" {
  algorithm = "RSA"
  rsa_bits  = 2048
}

output "ubuntu_vm_password" {
  value     = random_password.ubuntu_vm_password.result
  sensitive = true
}

output "ubuntu_vm_private_key" {
  value     = tls_private_key.ubuntu_vm_key.private_key_pem
  sensitive = true
}

output "ubuntu_vm_public_key" {
  value = tls_private_key.ubuntu_vm_key.public_key_openssh
}
```

**Comment by bpg-dev** [comment]
@NubemInvidia could you post an error you're getting?

Possibly related to #1599

**Reply by NubemInvidia** [reply]
Sure thing, sorry for the late reply:

```
│ Error: error waiting for VM creation: task "UPID:pve_cluster_node2:00306744:1700FABF:675F1355:qmcreate:800:svcterraform@pam!terraform:" failed to complete with exit code: unable to create VM 800 - can't allocate space in iscsi storage
│
│   with proxmox_virtual_environment_vm.ubuntu_vm,
│   on main.tf line 21, in resource "proxmox_virtual_environment_vm" "ubuntu_vm":
│   21: resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
│
╵
```

**Reply by NubemInvidia** [reply]
Still no luck figuring this out. [Documentation](https://registry.terraform.io/providers/bpg/proxmox/latest/docs/resources/virtual_environment_vm#size-2) seems to suggest that it will default to 8 gigabytes. This would make sense as to why it's erroring out during creation, since you can't allocate space on the iSCSI LUN. 

I could be missing something but this is what i've tried and this is the most recent error:
```  
disk {
    datastore_id = "testvm02"
    interface = "scsi0"
    #iothread = true
    aio = "native"
    backup = false
    cache = "none"
    file_format = "raw" #iscsi targets require this
    #size = 0
  }
```

resulting in:

```
│ Error: error waiting for VM creation: task "UPID:pve_cluster_node2:000017F6:29BBC912:678C6A87:qmcreate:804:svcterraform@pam!terraform:" failed to complete with exit code: unable to create VM 804 - can't allocate space in iscsi storage
│
│   with proxmox_virtual_environment_vm.ubuntu_vm,
│   on main.tf line 21, in resource "proxmox_virtual_environment_vm" "ubuntu_vm":
│   21: resource "proxmox_virtual_environment_vm" "ubuntu_vm" {
│
╵
```

**Comment by elias314** [comment]
It looks like Proxmox lets you use iSCSI LUNs in two different ways:

1.  Attach a LUN and then have Proxmox layer LVM over top of it so it behaves like any other storage.  When creating a disk for a VM, Proxmox will create an image file for it with the size you specify.  This appears to be the recommended approach per the [Proxmox documentation (see 7.13.3)](https://pve.proxmox.com/pve-docs/pve-admin-guide.html#_storage_features_9).

2. "Use the LUN directly" and attach it as a block device directly to the VM.  I suspect this is what you're trying to do.  I haven't tried codifying this in Terraform, but I did go through this in the Proxmox UI and this is what it looks like:

When adding the storage, click the "Use LUNs directly" box:
![Proxmox_Storage_iscsi](https://github.com/user-attachments/assets/483ed1d1-819c-49da-b90a-8341ebc49234)

With that box checked, the "size" field in the disk dialog gets replaced with a "Disk image" field, which auto-populated for me when I picked the LUN:
![Proxmox_iscsi_add_disk](https://github.com/user-attachments/assets/fac76b4c-ebbe-4c69-b325-a23afe2b0866)

Again, I haven't tried to use this via Terraform, nor do I know what's supported in the Proxmox API, but it looks like a different field is needed instead of "size" when directly attaching a LUN to a VM.

**Comment by NubemInvidia** [comment]
Ahh, I see. The solution to this is to create an LVM on top of a LUN instead of using the LUN directly as the whole disk for the VM. This way you can use the LVM as the datastore_id for the vm disk with terraform. 

For what it is worth, it does look like you can create an iscsi storage object with the storage api although i haven't been able to confirm it yet. Someone handier with APIs could probably confirm/deny.
